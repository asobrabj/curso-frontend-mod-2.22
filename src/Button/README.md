## Propriedades

O componente `Button` aceita as seguintes propriedades:

- **Children**
  - *Tipo:* `React.ReactNode`
  - *Descrição:* O conteúdo do botão.

- **as**
  - *Tipo:* `"h1" | "h2" | "h3" | "h4" | "h5" | "h6"`
  - *Descrição:* Define o tipo de elemento a ser renderizado. Útil para SEO e acessibilidade.

- **fs**
  - *Tipo:* `"xsmall" | "small" | "medium" | "larger" | string`
  - *Descrição:* Define o tamanho da fonte.

- **fStyle**
  - *Tipo:* `"normal" | "italic"`
  - *Descrição:* Define o estilo da fonte.

- **fw**
  - *Tipo:* `"medium" | "larger"`
  - *Descrição:* Define a espessura da fonte.

- **fFamily**
  - *Tipo:* `string`
  - *Descrição:* Define a família da fonte.

- **tAlign**
  - *Tipo:* `"start" | "end" | "justify" | "center"`
  - *Descrição:* Define o alinhamento do texto.

- **tTransform**
  - *Tipo:* `"none" | "capitalize" | "uppercase" | "lowercase"`
  - *Descrição:* Define a transformação do texto.

- **color**
  - *Tipo:* `string`
  - *Descrição:* Define a cor do texto.

- **bgColor**
  - *Tipo:* `string`
  - *Descrição:* Define a cor de fundo do botão.

- **pd, pdTop, pdRight, pdBottom, pdLeft**
  - *Tipo:* `"xsmall" | "small" | "medium" | "larger" | "xlarger" | "none"`
  - *Descrição:* Define o preenchimento (padding) do botão.

- **mg, mgTop, mgRight, mgBottom, mgLeft**
  - *Tipo:* `"xsmall" | "small" | "medium" | "larger" | "xlarger" | "none"`
  - *Descrição:* Define a margem do botão.

- **width, height**
  - *Tipo:* `string`
  - *Descrição:* Define a largura e altura do botão.

- **className**
  - *Tipo:* `string`
  - *Descrição:* Define classes adicionais para estilização personalizada.


## Classes do Componente Button

O componente `Button` oferece várias classes predefinidas para estilização personalizada. Abaixo estão as classes disponíveis:

### Classes para Estilos de Fundo e Texto

- **success**
  - *Descrição:* Aplica um estilo de sucesso ao botão, com texto branco e fundo verde.

- **info**
  - *Descrição:* Aplica um estilo de informação ao botão, com texto branco e fundo azul.

- **error**
  - *Descrição:* Aplica um estilo de erro ao botão, com texto branco e fundo vermelho.

- **alert**
  - *Descrição:* Aplica um estilo de alerta ao botão, com texto preto e fundo amarelo.

### Classes para Estilos de Borda

- **success-ligth**
  - *Descrição:* Aplica um estilo de borda leve ao botão de sucesso, com texto verde e borda verde.

- **info-ligth**
  - *Descrição:* Aplica um estilo de borda leve ao botão de informação, com texto azul e borda azul.

- **error-ligth**
  - *Descrição:* Aplica um estilo de borda leve ao botão de erro, com texto vermelho e borda vermelha.

- **alert-ligth**
  - *Descrição:* Aplica um estilo de borda leve ao botão de alerta, com texto amarelo, borda amarela e fundo cinza.

### Classes para Estilos Especiais

- **rounded**
  - *Descrição:* Aplica uma borda arredondada de 8 pixels ao botão.

- **circle**
  - *Descrição:* Transforma o botão em um círculo com 50 pixels de largura, 50 pixels de altura e borda arredondada.

    `


