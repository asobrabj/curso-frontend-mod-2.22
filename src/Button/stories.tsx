import Button, { IButton } from ".";

export default {
  title: "Components/Button",
  component: Button,
  args: {
    children: "Hello world",
  },
};

export const template = (args: IButton) => (
  <div>
    <Button className="alert-ligth" {...args} />
  </div>
);
