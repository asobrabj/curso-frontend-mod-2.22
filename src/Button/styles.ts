import styled from 'styled-components';
import {
  IStyle,
  fontSize,
  fontWheigt,
  marginFn,
  paddingFn,
} from '../../types/styles';

export const Button = styled.button<IStyle>`
  font-size: ${(props) => (props.$fs ? fontSize(props.$fs) : '')};
  font-style: ${(props) => props.$fStyle || ''};
  font-weight: ${(props) => (props.$fw ? fontWheigt(props.$fw) : '')};
  font-family: ${(props) => props.$fFamily || ''};
  text-align: ${(props) => props.$tAlign || ''};
  text-transform: ${(props) => props.$tTransform || ''};
  color: ${(props) => props.$color || ''};
  background-color: ${(props) => props.$bgColor || ''};
  padding: ${(props) => (props.$pd ? paddingFn(props.$pd) : '')};
  padding-top: ${(props) => (props.$pdTop ? props.$pdTop + 'px' : '')};
  padding-left: ${(props) => (props.$pdLeft ? props.$pdLeft + 'px' : '')};
  padding-right: ${(props) => (props.$pdRigth ? props.$pdRigth + 'px' : '')};
  padding-bottom: ${(props) => (props.$pdBottom ? props.$pdBottom + 'px' : '')};
  margin: ${(props) => (props.$mg ? marginFn(props.$mg) : '')};
  margin-top: ${(props) => (props.$mgTop ? props.$mgTop + 'px' : '')};
  margin-left: ${(props) => (props.$mgLeft ? props.$mgLeft + 'px' : '')};
  margin-right: ${(props) => (props.$mgRigth ? props.$mgRigth + 'px' : '')};
  margin-bottom: ${(props) => (props.$mgBottom ? props.$mgBottom + 'px' : '')};
  width: ${(props) => (props.$width ? props.$width + 'px' : '150')};
  height: ${(props) => (props.$height ? props.$height + 'px' : '')};
  transition:
    box-shadow 0.3s ease,
    transform 0.3s ease;

  &.success,
  &.success-ligth,
  &.info,
  &.info-ligth,
  &.error,
  &.error-ligth,
  &.alert,
  &.alert-ligth {
    font-weight: ${(props) => (props.$fw ? fontWheigt(props.$fw) : 'bold')};
    text-align: ${(props) => props.$tAlign || 'center'};
    border-radius: 16px;
    box-shadow: 2px 2px 2px rgba(0, 0, 0, 0.5);
    border: none;
  }

  &.success {
    color: #ffffff;
    background-color: #83d563;
  }

  &.info {
    color: #ffffff;
    background-color: #00abff;
  }

  &.error {
    color: #ffffff;
    background-color: #ff0000;
  }

  &.alert {
    color: #000000;
    background-color: #f1ea13;
  }

  &.success-ligth {
    color: #83d563;
    border: 2px solid #83d563;
  }

  &.info-ligth {
    color: #00abff;
    border: 2px solid #00abff;
  }

  &.error-ligth {
    color: #ff0000;
    border: 2px solid #ff0000;
  }

  &.alert-ligth {
    color: #f1ea13;
    border: 2px solid #f1ea13;
    background-color: #8d877a;
  }

  &.rounded {
    border-radius: 8px;
  }

  &.circle {
    width: 50px;
    height: 50px;
    border-radius: 50%;
  }

  &:hover {
    cursor: pointer;
  }

  &:active {
    box-shadow: 0 2px 2px rgba(0, 0, 0, 0.2);
    transform: translatey(2px);
  }
`;
