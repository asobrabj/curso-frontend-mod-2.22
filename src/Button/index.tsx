import { GlobalProps } from '../../types';
import * as S from './styles';

export interface IButton extends GlobalProps {
  children: React.ReactNode;
  onClick?: () => void;
}

export default function Button(props: IButton) {
  return (
    <S.Button
      onClick={props.onClick}
      type="submit"
      className={props.className}
      $fs={props.fs}
      $fw={props.fw}
      $fStyle={props.fStyle}
      $fFamily={props.fFamily}
      $tAlign={props.tAlign}
      $tTransform={props.tTransform}
      $color={props.color}
      $bgColor={props.bgColor}
      $pd={props.pd}
      $pdTop={props.pdTop}
      $pdRigth={props.pdRigth}
      $pdLeft={props.pdLeft}
      $pdBottom={props.pdBottom}
      $mg={props.mg}
      $mgTop={props.mgTop}
      $mgRigth={props.mgRigth}
      $mgLeft={props.mgLeft}
      $mgBottom={props.mgBottom}
      $width={props.width}
      $height={props.height}
    >
      {props.children}
    </S.Button>
  );
}
