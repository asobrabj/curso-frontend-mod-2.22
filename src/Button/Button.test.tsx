import { render, screen } from '@testing-library/react';
import Button from './index';

describe('Button Component', () => {
  it('renders button text correctly', () => {
    const { getByText } = render(<Button>Click me</Button>);
    const buttonElement = getByText(/Click me/i);
    expect(buttonElement).toBeInTheDocument();
  });

  it('applies font-size correctly', () => {
    render(<Button fs="50">Styled Button</Button>);
    const buttonElement = screen.getByText(/Styled Button/i);
    expect(buttonElement).toHaveStyle('font-size:50px');
    expect(buttonElement).not.toHaveStyle('font-size:51px');
  });

  it('renders with margin correcty', () => {
    render(<Button mg="medium">Default Button</Button>);
    const buttonElement = screen.getByText(/Default Button/i);
    expect(buttonElement).toHaveStyle('margin: 16px');
    expect(buttonElement).not.toHaveStyle('margin: 17px');
  });

  it('renders with padding correcty', () => {
    render(<Button pd="medium">Default Button</Button>);
    const buttonElement = screen.getByText(/Default Button/i);
    expect(buttonElement).toHaveStyle('padding: 16px');
    expect(buttonElement).not.toHaveStyle('padding: 14px');
  });

  it('renders with padding correcty', () => {
    render(
      <Button
        pd="medium"
        fStyle="italic"
        fw="larger"
        height="30"
        tAlign="center"
        width="150"
      >
        Default Button
      </Button>,
    );
    const buttonElement = screen.getByText(/Default Button/i);
    expect(buttonElement).toHaveStyle('padding: 16px');
    expect(buttonElement).not.toHaveStyle('padding: 14px');
    expect(buttonElement).toHaveStyle('font-style: italic');
    expect(buttonElement).not.toHaveStyle('font-style: normal');
    expect(buttonElement).toHaveStyle('font-weight: bold');
    expect(buttonElement).not.toHaveStyle('font-weight: 200');
    expect(buttonElement).toHaveStyle('height: 30px');
    expect(buttonElement).not.toHaveStyle('height: 32px');
    expect(buttonElement).toHaveStyle('text-align: center');
    expect(buttonElement).not.toHaveStyle('text-align: end');
    expect(buttonElement).toHaveStyle('width: 150px');
    expect(buttonElement).not.toHaveStyle('width: 155px');
  });

  it('match snapshot', () => {
    render(<Button pd="medium">Default Button</Button>);
    const buttonElement = screen.getByText(/Default Button/i);
    expect(buttonElement).toMatchSnapshot();
  });
});
