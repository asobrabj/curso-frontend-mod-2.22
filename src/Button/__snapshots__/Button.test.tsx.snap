// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component match snapshot 1`] = `
<button
  class="sc-bdfCDU cLweCQ"
  type="submit"
>
  Default Button
</button>
`;
